feat(positioning): add positioning service

RobJacobs opened the pull request:
Closes #289

Need feedback/help with::

1.  Is the name 'Positioning' acceptable?
2.  Does the Positioning class need to be marked as @Injectable?
3.  I'm getting conflicting clang-format and tslint errors on spaces around curly braces, tslint wants:

` func(): number { return 0; }`

clang-format wants:

` func(): number{return 0;}`

to get the build to complete locally, I had to remove the tslint rules.
4.  Rather than create an ELementPosition interface, should I use the ClientRect interface instead?  It has 2 properties (right and bottom) that we don't currently return, but we could easily enough. 

I still need to write the test, but I wanted to make sure I'm going down the right path, thanks.

icfantv commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,102 @@
+      parentOffset.left += offsetParentEl.clientLeft - offsetParentEl.scrollLeft;
+    }
+
+    return <ElementPosition>{
```
don't need the cast here.

RobJacobs replied:
It adds type safety to the object I'm returning, or is that not important?

icfantv replied:
the compiler will check that if the function declares a return type, that the returned type matches what's declared.

icfantv commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,102 @@
+
+  offset(element: HTMLElement): ElementPosition {
+    let elBcr: ClientRect = element.getBoundingClientRect();
+    return <ElementPosition>{
```
ditto

icfantv commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,102 @@
+import {ElementRef} from '@angular/core';
+
+export interface ElementPosition {
```
should be named `IElementPosition` since it is an interface

pkozlowski-opensource commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,102 @@
+import {ElementRef} from '@angular/core';
```
Doesn't look like we are using `ElementRef` anywhere (and probably shouldn't be) so this import can be removed.

pkozlowski-opensource commented:
@RobJacobs I will review in grater details tomorrow but I was wondering if there is any way to write meaningful tests for this service?

RobJacobs commented:
@pkozlowski-opensource I'm going to look at how material is testing their stuff and see what we can do.

wesleycho commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,100 @@
+    let elOffset: ElementPosition = this.offset(element);
+    let elBCR: ClientRect = element.getBoundingClientRect();
+    let parentOffset: ElementPosition = {width: 0, height: 0, top: 0, left: 0};
+    let offsetParentEl: HTMLElement = this.parentOffsetElement(element);
```
The typings in these several of lines are unnecessary I think

wesleycho replied:
Well, excepting `parentOffset` - that is probably ok.

wesleycho commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,100 @@
+    let shiftWidth: any = {
+      left(): number{return hostElPosition.left;},
+      center(): number{return hostElPosition.left + hostElPosition.width / 2 - targetElement.offsetWidth / 2;},
+      right(): number{return hostElPosition.left + hostElPosition.width;}
```
How come these have to be functions? Is this to readjust after the element has been rendered?

wesleycho commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,100 @@
+    let shiftHeight: any = {
+      top(): number{return hostElPosition.top;},
+      center(): number{return hostElPosition.top + hostElPosition.height / 2 - targetElement.offsetHeight / 2;},
+      bottom(): number{return hostElPosition.top + hostElPosition.height;}
```
Same question as above

wesleycho commented on src/util/positioning.ts:
```diff
@@ -0,0 +1,100 @@
+      center(): number{return hostElPosition.top + hostElPosition.height / 2 - targetElement.offsetHeight / 2;},
+      bottom(): number{return hostElPosition.top + hostElPosition.height;}
+    };
+    let targetElBCR: ClientRect = targetElement.getBoundingClientRect();
```
Typing can probably be removed as inferred.

pkozlowski-opensource commented:
@RobJacobs as a very minimum we should have tests for calculations (those could be decoupled from the DOM - that is we could write tests for the part that does calculations after dimensions of all the involved elements is known).

RobJacobs commented:
Updated based on suggestions.  I switched from creating an ElementPosition interface to using the Angular ClientRect interface and now return right/bottom values.  That will put us in a better position to support smart positioning down the road.  I'm going to work on tests today.

RobJacobs commented:
@pkozlowski-opensource Tests added, did a little more refactoring.  I think this is good to go, let me know if you see anything that needs work.

wesleycho commented on src/util/positioning.spec.ts:
```diff
@@ -0,0 +1,263 @@
+    element.style.height = height + 'px';
+    element.style.width = width + 'px';
+    element.style.marginTop = marginTop + 'px';
+    element.style.marginLeft = marginLeft + 'px';
```
Could use template strings for these 4 lines

pkozlowski-opensource commented:
@RobJacobs thnx, I've merged it to unblock the situation. We know that we need to improve it but I think it is good as a starting point. Expect to hear from me more as soon as I start impl of tooltips / popovers (just after alpha.0 is out).
